prometheus/reporter.go

```diff
@@ -0,0 +1,254 @@
+}
+
+// RegisterGauge is a helper method to initialize a gauge in the prometheus backend with a given help text.
+// If not called explicitly, the Reporter will create one for you on first use, with a not super helpful HELP string
+func (r *Reporter) RegisterGauge(name string, tags map[string]string, desc string) (*prom.GaugeVec, error) {
+	g := &prom.GaugeVec{}
+	id := hashMetricLabelsToID(name, tags)
+	exists := r.hasGauge(id)
+	if exists {
+		return g, errorAlreadyRegistered
+	}
+	labelKeys := keysFromMap(tags)
+
+	g = prom.NewGaugeVec(
+		prom.GaugeOpts{
+			Name: name,
+			Help: desc,
+		},
+		labelKeys,
+	)
+	err := prom.Register(g)
+	if err != nil {
+		return g, err
+	}
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	r.gauges[id] = g
+	return g, nil
+}
+
+// ReportGauge reports a gauge value
+func (r *Reporter) ReportGauge(name string, tags map[string]string, value float64) {
+	id := hashMetricLabelsToID(name, tags)
+
+	r.mtx.RLock()
+	g, ok := r.gauges[id]
+	r.mtx.RUnlock()
+
+	if !ok {
+		var err error
+		g, err = r.RegisterGauge(name, tags, name+" gauge")
+		if err != nil {
+			panic(err)
+		}
+	}
+	g.With(tags).Set(value)
+}
+
+// RegisterTimer is a helper method to initialize a Timer histogram vector in the prometheus backend with a given help text.
+// If not called explicitly, the Reporter will create one for you on first use, with a not super helpful HELP string
+// objectives is the prometheus.SummaryVec objectives. Default, if nil passed is {0.5: 0.05, 0.9: 0.01, 0.99: 0.001, 0.999: 0.0001}
+// See https://godoc.org/github.com/prometheus/client_golang/prometheus#SummaryOpts
+func (r *Reporter) RegisterTimer(name string, tags map[string]string, desc string, objectives map[float64]float64) (*prom.SummaryVec, error) {
+	h := &prom.SummaryVec{}
+	id := hashMetricLabelsToID(name, tags)
+	exists := r.hasSummary(id)
+	if exists {
+		return h, errorAlreadyRegistered
+	}
+	labelKeys := keysFromMap(tags)
+
+	if objectives == nil {
+		objectives = map[float64]float64{0.5: 0.05, 0.9: 0.01, 0.99: 0.001, 0.999: 0.0001}
+	}
+	h = prom.NewSummaryVec(
+		prom.SummaryOpts{
+			Name:       name,
+			Help:       desc,
+			Objectives: objectives,
+		},
+		labelKeys,
+	)
+	err := prom.Register(h)
+	if err != nil {
+		return h, err
+	}
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	r.summaries[id] = h
+	return h, nil
+}
+
+// ReportTimer reports a timer value into the Summary histogram
+func (r *Reporter) ReportTimer(name string, tags map[string]string, interval time.Duration) {
+	id := hashMetricLabelsToID(name, tags)
+
+	r.mtx.RLock()
+	h, ok := r.summaries[id]
+	r.mtx.RUnlock()
+
+	if !ok {
+		var err error
+		h, err = r.RegisterTimer(name, tags, name+" histogram in seconds", nil)
+		if err != nil {
+			panic(err)
+		}
+	}
+	h.With(tags).Observe(float64(interval))
+}
+
+// Capabilities ...
+func (r *Reporter) Capabilities() tally.Capabilities {
+	return r
+}
+
+// Reporting ...
+func (r *Reporter) Reporting() bool {
+	return false
```

Comment:
Reporting indicates it can report outside of the process.  Since its prometheus and can be scraped from this should probably be `true`.